common/src/main/java/io/pravega/common/io/filesystem/FileModificationPollingMonitor.java

```diff
@@ -0,0 +1,138 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.io.filesystem;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.monitor.FileAlterationListener;
+import org.apache.commons.io.monitor.FileAlterationListenerAdaptor;
+import org.apache.commons.io.monitor.FileAlterationMonitor;
+import org.apache.commons.io.monitor.FileAlterationObserver;
+
+import java.io.File;
+import java.io.FileFilter;
+import java.io.FileNotFoundException;
+import java.nio.file.InvalidPathException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.function.Consumer;
+
+/**
+ * Monitors for modifications to the specified file using a polling-based mechanism and performs the
+ * specified action (in the form of a callback) upon detection of a file modification.
+ *
+ * Unlike {@link FileModificationEventWatcher}, which doesn't work for monitoring modifications to a file which is
+ * a symbolic link to another file, an object of this class works for such files as well. However, since this object
+ * uses a polling-based mechanism - unlike the event-based mechanism used by {@link FileModificationEventWatcher}, lists
+ * files on each iteration and compares the output, this object is highly likely to be less efficient than an object of
+ * {@link FileModificationEventWatcher} - CPU cycles-wise.
+ *
+ */
+@Slf4j
+public class FileModificationPollingMonitor implements FileModificationMonitor {
+
+    private final static int DEFAULT_POLL_INTERVAL = 10 * 1000; // 10 seconds
+
+    /**
+     * The path of file to watch.
+     */
+    private final Path pathOfFileToWatch;
+
+    /**
+     * The action to perform when the a modification is detected for the specified file {@code pathOfFileToWatch}.
+     */
+    private final Consumer<File> callback;
+
+    private FileAlterationMonitor monitor;
+
+    private int pollingInterval;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param fileToWatch the file to watch
+     * @param callback    the callback to invoke when a modification to the {@code fileToWatch} is detected
+     *
+     * @throws NullPointerException if either {@code fileToWatch} or {@code callback} is null
+     * @throws InvalidPathException if {@code fileToWatch} is invalid
```

Comment:
Where is this being thrown from?

Reply:
`Paths.get(..)` may throw `InvalidPathException`, as described [here](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Paths.html#get-java.lang.String-java.lang.String...-).